lec08-stability.Rmd

```diff
@@ -1,4 +1,629 @@
+mass_spring_trajectory <- trajectory(mass_spring, 
+                            # y0 is a matrix where each row is pairs of (X, Y) 
+                            y0 = matrix(c(1, 1, 0, 1, 2, -1.5), ncol = 2,
+                                 byrow = TRUE), 
+                            t.end = 20, # how far in time to calculate the trajectories
+                            parameters = c(omega = 0.5), 
+                            system = "two.dim")
+```
+
+## Finding fixed points analytically (time permitting)
+
+Up until now we've been plotting `r eq_dn_dt` vs. $N$ to find the fixed points of a system. 
+But we can also solve `r eq_dn_dt` $=0$ analytically to find the fixed points. One advantage of
+finding the fixed points analytically is that you can see how the fixed points depend on the
+model parameters, whereas before we had to choose parameters in order to plot `r eq_dn_dt`
+and find the fixed points qualitatively.
+
+$$\frac{dN}{dt} = rN(1-\frac{N}{K})$$
+
+First, we set `r eq_dn_dt` $=0$. 
+
+$$0 = rN(1-\frac{N}{K})$$
+
+Next, we find all the values of $N$ that satisfy this equation. If either $rN = 0$ or 
+$1-\frac{N}{K} = 0$, the equation is satisfied. This means we have two fixed points, 
+one at $N = 0$ (since $r \neq 0$) and the other at $N = K$. These are the same fixed 
+points we found graphically last week. Notice that one of the fixed points depends on 
+$K$, but that neither of the fixed points depend on $r$. 
+
+#### Challenge
+
+Find the fixed point(s) of the following differential equation: 
+
+$$\frac{dx}{dt} = x^2 - 1$$
+
+   - First find the fixed points analytically on paper.
+   - Plot $\frac{dx}{dt}$ vs. $x$ for $x$ ranging from $-2$ to $2$ in $0.1$ increments.
+   - Sketch a phase portrait. Are the fixed point(s) stable or unstable?
+
+```{r, echo=FALSE, eval=FALSE}
+# Challenge solution
+
+# fixed points are x = 1, x = -1
+
+library(ggplot2)
+
+x_seq <- seq(-2, 2, by = 0.1)
+
+dx_dt <- x_seq^2 - 1
+
+qplot(x_seq, dx_dt) +
+  geom_hline(aes(yintercept = 0))
+```
+
+## Extras
+
+Predator-prey simulation GIF in R
+
+```{r, eval=FALSE}
+library(animation)
+library(ggplot2)
+library(grid)
+library(gridExtra)
+
+ani.options(interval = 0.00001)
+
+parameters = c(a = 5, b = 1, c = 1, d = 0.2) 
+state <- c(X = 2, Y = 2)
+times <- seq(0, 50, by = 0.01)
+
+result <- ode(y = state, times = times, func = predator_prey, parms = parameters)
+result <- data.frame(result)
+
+saveGIF({ 
+    for (i in seq(1, length(result$time), by = 50)) { 
+      result_i <- head(result, i)
+      p1 <- ggplot(result_i) +
+        geom_line(aes(x = result_i$time, y = result_i$X), color = 'red') +
+        geom_line(aes(x = result_i$time, y = result_i$Y), color = 'blue') +
+        ylim(0, 8) +
+        xlab("Time") +
+        ylab("Population size")
+
+      p2 <- ggplot(result_i) +
+        geom_path(aes(x = result_i$X, y = result_i$Y, color = result_i$time)) +
+        guides(color = guide_legend("Time")) + 
```

Comment:
Is there a way to fix the scale for the colour legend so that it doesn't keep changing in the GIF? I couldn't figure that out.

Reply:
Tagging @lwjohnst86

Reply:
I'll look at it. Also, you don't need to use `result_i$X` inside ggplot. `X` will do.

Reply:
I looked into a bit, and I know there's a way to fix that issue, but I don't readily know how without doing some digging ... and if I had time, I would dig more, but alas I don't have a lot of time :(

A easy solution would be to remove the legend entirely.

Reply:
Remove legend: `theme(legend.position="none")`